sdks/python/apache_beam/dataframe/doctests.py

```diff
@@ -428,7 +482,106 @@ def print_partition(indent, desc, n, total):
     print()
 
 
-def teststring(text, report=True, **runner_kwargs):
+def parse_rst_ipython_tests(rst, name, extraglobs=None, optionflags=None):
+  """Extracts examples from an rst file and produce a test suite by running
+  them through pandas to get the expected outputs.
+  """
+
+  # Optional dependency.
+  import IPython
+  from traitlets.config import Config
+
+  def get_indent(line):
+    return len(line) - len(line.lstrip())
+
+  def is_example_line(line):
+    line = line.strip()
+    return line and not line.startswith('#') and not line[0] == line[-1] == ':'
+
+  IMPORT_PANDAS = 'import pandas as pd'
+
+  example_srcs = []
+  lines = iter(
+      [line.rstrip()
+       for line in rst.split('\n') if is_example_line(line)] + ['END'])
+
+  # https://ipython.readthedocs.io/en/stable/sphinxext.html
+  for line in lines:
+    if line.startswith('.. ipython::'):
+      line = next(lines)
+      indent = get_indent(line)
+      example = []
+      example_srcs.append(example)
+      while get_indent(line) >= indent:
+        if '@verbatim' in line or '@savefig' in line:
```

Comment:
We should probably check for `:verbatim:` https://ipython.readthedocs.io/en/stable/sphinxext.html#pseudo-decorators